tuta-sdk/rust/sdk/src/lib.rs

```diff
@@ -90,88 +90,90 @@ impl Display for TypeRef {
-    client_version: String,
+    fn provide_headers(&self, model_version: u32) -> HashMap<String, String> {
+        HashMap::from([
+            ("accessToken".to_string(), self.access_token.clone()),
+            ("cv".to_owned(), self.client_version.clone()),
+            ("v".to_owned(), model_version.to_string()),
+        ])
+    }
 }
 
 /// The external facing interface used by the consuming code via FFI
 #[derive(uniffi::Object)]
 pub struct Sdk {
-    state: Arc<SdkState>,
     type_model_provider: Arc<TypeModelProvider>,
-    entity_client: Arc<EntityClient>,
-    typed_entity_client: Arc<TypedEntityClient>,
+    json_serializer: Arc<JsonSerializer>,
     instance_mapper: Arc<InstanceMapper>,
+    rest_client: Arc<dyn RestClient>,
+    base_url: String,
+    client_version: String,
 }
 
 #[uniffi::export]
 impl Sdk {
     #[uniffi::constructor]
-    pub fn new(base_url: String, rest_client: Arc<dyn RestClient>, credentials: Credentials, client_version: &str) -> Sdk {
+    pub fn new(base_url: String, rest_client: Arc<dyn RestClient>, client_version: String) -> Sdk {
         logging::init_logger();
         log::debug!("Initializing SDK...");
 
         let type_model_provider = Arc::new(init_type_model_provider());
         // TODO validate parameters
-        let json_serializer = Arc::new(JsonSerializer::new(type_model_provider.clone()));
         let instance_mapper = Arc::new(InstanceMapper::new());
-        let state = Arc::new(SdkState {
-            credentials,
-            client_version: client_version.to_owned(),
-        });
-        let entity_client = Arc::new(EntityClient::new(
-            rest_client,
-            json_serializer,
-            &base_url,
-            state.clone(),
-            type_model_provider.clone(),
-        ));
-        let typed_entity_client: Arc<TypedEntityClient> = Arc::new(TypedEntityClient::new(
-            entity_client.clone(),
-            instance_mapper.clone()
-        ));
+        let json_serializer = Arc::new(JsonSerializer::new(type_model_provider.clone()));
 
         Sdk {
-            state,
             type_model_provider,
-            entity_client,
-            typed_entity_client,
-            instance_mapper
+            json_serializer,
+            instance_mapper,
+            rest_client,
+            base_url,
+            client_version,
         }
     }
 
     /// Authorizes the SDK's REST requests via inserting `access_token` into the HTTP headers
-    pub async fn login(&self) -> Result<Arc<LoggedInSdk>, LoginError> {
+    pub async fn login(&self, credentials: Credentials) -> Result<Arc<LoggedInSdk>, LoginError> {
+        let auth_headers_provider = Arc::new(HeadersProvider::new(
+            self.client_version.clone(),
+            credentials.access_token.clone(),
+        ));
+        let entity_client = Arc::new(EntityClient::new(
+            self.rest_client.clone(),
+            self.json_serializer.clone(),
+            self.base_url.clone(),
```

Comment:
Perhaps pass this by reference?

Reply:
Would be great! but we would need to define the lifetime relationship between `Sdk` and `EntityClient` and I think Uniffi types are not allowed to have lifetimes.